### PR TITLE
Integrate BMO V4 Fuel Cost Model and Dry-Basis Blend Calculations

### PR DESCRIPTION
**Pull request by charan-x16**
# PR Description

## Overview

This PR enhances the Blend Mix Optimizer (BMO) by integrating the latest BMO V4 fuel-cost prediction model and updating blend evaluation calculations to follow dry-basis metallurgical formulas.

The implementation aligns optimizer behavior with the latest model training pipeline and blast furnace burden calculation methodology, resulting in more accurate Fe%, slag, and total-cost evaluation.

A separate code-change document is attached for detailed implementation flow, file-level changes, and function-level explanations.
[BMO_CHANGED_CODE_EXPLANATION.pdf](https://github.com/user-attachments/files/28408674/BMO_CHANGED_CODE_EXPLANATION.pdf)

## Why This Change Was Needed

The existing Blend Mix Optimizer required updates in two key areas:

### Fuel Cost Prediction

* Existing fuel-cost prediction logic was not aligned with the latest trained BMO V4 model.
* Legacy model-loading approaches could generate XGBoost and scikit-learn compatibility warnings.
* Feature generation did not fully mirror the model training pipeline.

### Blend Evaluation Logic

* Fe% calculations did not consistently use dry-weight weighted averaging.
* Slag calculations did not use dry-weight oxide contribution formulas.
* The optimizer relied on artificial total burden quantity constraints that did not always reflect realistic blast furnace operating conditions.
* Duplicate BMO implementations increased maintenance complexity and risk of inconsistent behavior.

These changes improve blend accuracy, model reliability, and optimizer maintainability while producing results that are easier to validate against plant and DPR data.

### BMO V4 Fuel Model Integration

* Integrated the latest BMO V4 model bundle:

  * `bmo_xgb_model.json`
  * `bmo_scaler.joblib`
  * `bmo_feature_columns.json`
* Added model bundle loading through runtime services.
* Implemented feature-building logic that matches the model training schema.
* Added support for selected-feature inference flow.
* Added fallback prediction logic when model assets are unavailable.

### Fuel Cost Prediction

* Added fuel-cost prediction to blend evaluation.
* Added fuel-cost enrichment for LP baseline results.
* Integrated fuel-cost prediction into Differential Evolution optimization.
* Included fuel cost in total objective calculation.

### Blend Calculation Enhancements

* Updated blend calculations to use dry-weight burden methodology.
* Added moisture-aware dry-weight calculations.
* Updated Fe contribution calculations.
* Updated final Fe% calculation using dry-weight weighted averaging.
* Updated slag calculations using dry-weight slag-forming oxide contributions.

### Optimizer Enhancements

* Removed fixed total quantity constraints.
* Removed artificial maximum total burden quantity caps.
* Updated LP constraints to use dry-weight Fe and slag coefficients.
* Updated DE optimization to evaluate realistic burden quantities based on:

  * Fe production targets
  * Slag limits
  * Material stock limits
  * Share constraints

### UI Improvements

Updated Blend Optimizer dashboard metrics to display:

* Objective cost
* Ore cost
* Fuel cost
* Fe production
* Wet quantity
* Dry quantity
* Final Fe%
* Effective Fe%
* Slag MT
* Slag rate
* Material-level Fe contribution
* Material-level slag contribution

### Data and Configuration Updates

* Added moisture mapping support.
* Added Na₂O mapping support.
* Added K₂O mapping support.
* Updated BMO model configuration to use V4 assets.
* Improved online chemistry and stock integration.
* Added fallback handling for unavailable online data.

### Code Consolidation

* Consolidated active BMO implementation under:

```text
src/utils/bmo
```

* Removed duplicate implementation under:

```text
src/domain/bmo
```

* Updated imports and runtime references accordingly.

### Testing

Added and updated test coverage for:

* BMO context provider
* Fuel model integration
* Feature generation
* Dry-weight Fe calculations
* Dry-weight slag calculations
* LP optimization flow
* DE optimization flow
* Constraint validation
* Runtime model services

### Dry Weight

```text
Dry Weight =
Wet Weight × (100 − Moisture%) / 100
```

### Fe Contribution

```text
Fe MT =
Dry Weight × Fe% / 100
```

### Final Fe%

```text
Final Fe% =
Σ(Fe MT) / Σ(Dry Weight) × 100
```

### Slag Oxide Sum

```text
Oxide Sum =
SiO2 + Al2O3 + CaO + MgO +
TiO2 + MnO + Na2O + K2O
```

### Slag Contribution

```text
Slag MT =
Dry Weight × Oxide Sum / 100
```

### Slag Rate

```text
Slag Rate (kg/THM) =
Slag MT / Fe Production MT × 1000
```

### Code Quality Checks

```bash
git diff --check
```

Result:

* No whitespace issues found.
* No merge-conflict markers found.

### Automated Tests

```bash
uv run pytest \
tests/test_bmo_context_provider.py \
tests/test_bmo_v4_integration.py \
tests/test_optimization_runtime.py -q
```

Result:

```text
18 passed
```

## Key Outcomes

* BMO now uses the latest V4 fuel-cost prediction model.
* Blend calculations follow dry-basis metallurgical formulas.
* Fe% and slag calculations are more accurate and traceable.
* Optimizer recommends realistic burden quantities based on operational constraints.
* Duplicate BMO code paths have been removed.
* LP and DE optimization flows remain fully operational with updated logic.
* Improved maintainability, validation, and alignment with plant operating practices.

### CODE REVIEW

**Comment by charan-x16** [comment]
Evidence attachment includes the Blend Mix Optimizer dashboard screenshots and supporting code-change documentation. The screenshots show the updated LP Baseline, Total Cost Optimizer, comparison outputs, dry quantity, Fe production, final Fe%, slag MT, slag rate, and material-level blend details. The attached documentation explains the changed files, function flow, model integration, and dry-basis Fe/slag calculation logic used in this implementation.
[BMO-Evidence.docx](https://github.com/user-attachments/files/28408776/BMO-Evidence.docx)
